File: src/lib/schemas/pack.ts
import { Schema } from 'effect';

import { GamemodeId, License, LocaleCode, PackId, QuestionId, TagRef } from './common.ts';
import { QuestionKind, VariantName } from './question.ts';

export const PackFilter = Schema.Struct({
	kinds: Schema.optionalKey(Schema.Array(QuestionKind)),
	limit: Schema.optionalKey(Schema.Number.check(Schema.isInt(), Schema.isGreaterThanOrEqualTo(1))),
	tags_all: Schema.optionalKey(Schema.Array(TagRef)),
	tags_any: Schema.optionalKey(Schema.Array(TagRef)),
	tags_none: Schema.optionalKey(Schema.Array(TagRef)),
	variants_any: Schema.optionalKey(Schema.Array(VariantName))
});
export type PackFilter = typeof PackFilter.Type;

export const Pack = Schema.Struct({
	author: Schema.optionalKey(Schema.String),
	default_lang: Schema.optionalKey(LocaleCode),
	description: Schema.optionalKey(Schema.String),
	filter: Schema.optionalKey(PackFilter),
	id: PackId,
	includes: Schema.optionalKey(Schema.Array(PackId)),
	license: Schema.optionalKey(License),
	questions: Schema.optionalKey(Schema.Array(QuestionId)),
	recommended_gamemodes: Schema.optionalKey(Schema.Array(GamemodeId)),
	title: Schema.String
}).check(
	Schema.makeFilter((p) =>
		(p.includes !== undefined && p.includes.length > 0) ||
		(p.questions !== undefined && p.questions.length > 0) ||
		p.filter !== undefined
			? undefined
			: 'pack must define at least one of: includes, questions, filter'
	)
);
export type Pack = typeof Pack.Type;

// Pack files are single-pack-per-file (unlike question files which are arrays).
export const PackFile = Pack;
export type PackFile = typeof PackFile.Type;

export const PackOverlay = Schema.Struct({
	description: Schema.optionalKey(Schema.String),
	id: PackId,
	title: Schema.optionalKey(Schema.String)
});
export type PackOverlay = typeof PackOverlay.Type;

export const PackOverlayFile = PackOverlay;
export type PackOverlayFile = typeof PackOverlayFile.Type;

File: api/src/data/types/pack.rs
use garde::Validate;
use serde::Deserialize;

use super::common::*;

#[derive(Debug, Clone, Deserialize, Validate)]
#[garde(allow_unvalidated)]
pub struct PackFilter {
    #[serde(default)]
    pub kinds: Option<Vec<QuestionKind>>,
    #[serde(default)]
    pub tags_all: Option<Vec<String>>,
    #[serde(default)]
    pub tags_any: Option<Vec<String>>,
    #[serde(default)]
    pub tags_none: Option<Vec<String>>,
    #[serde(default)]
    pub variants_any: Option<Vec<VariantName>>,
    #[serde(default)]
    pub limit: Option<usize>,
}

#[derive(Debug, Clone, Deserialize, Validate)]
#[garde(allow_unvalidated)]
#[garde(custom(pack_has_content))]
pub struct Pack {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub author: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub license: Option<License>,
    #[serde(default)]
    pub default_lang: Option<String>,
    #[serde(default)]
    pub recommended_gamemodes: Option<Vec<String>>,
    #[serde(default)]
    pub includes: Option<Vec<String>>,
    #[serde(default)]
    pub questions: Option<Vec<String>>,
    #[garde(dive)]
    #[serde(default)]
    pub filter: Option<PackFilter>,
}

fn pack_has_content(pack: &Pack, _ctx: &()) -> garde::Result {
    let has_includes = pack.includes.as_ref().is_some_and(|v| !v.is_empty());
    let has_questions = pack.questions.as_ref().is_some_and(|v| !v.is_empty());
    let has_filter = pack.filter.is_some();
    if has_includes || has_questions || has_filter {
        Ok(())
    } else {
        Err(garde::Error::new(
            "pack must define at least one of: includes, questions, filter",
        ))
    }
}

#[derive(Debug, Clone, Deserialize, Validate)]
#[garde(allow_unvalidated)]
pub struct PackOverlay {
    pub id: String,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
}
